packages/testharness/src/speech/speechRecognition/createQueuedArrayBufferAudioSource.js

```diff
@@ -0,0 +1,160 @@
+import { AudioStreamFormat } from '../../external/microsoft-cognitiveservices-speech-sdk';
+
+import {
+  AudioSourceErrorEvent,
+  AudioSourceInitializingEvent,
+  AudioSourceOffEvent,
+  AudioSourceReadyEvent,
+  AudioStreamNodeAttachedEvent,
+  AudioStreamNodeAttachingEvent,
+  AudioStreamNodeDetachedEvent
+  // AudioStreamNodeErrorEvent,
+} from '../../external/microsoft-cognitiveservices-speech-sdk/distrib/lib/src/common/AudioSourceEvents';
+
+import { createNoDashGuid } from '../../external/microsoft-cognitiveservices-speech-sdk/distrib/lib/src/common/Guid';
+import { Events } from '../../external/microsoft-cognitiveservices-speech-sdk/distrib/lib/src/common/Events';
+import { EventSource } from '../../external/microsoft-cognitiveservices-speech-sdk/distrib/lib/src/common/EventSource';
+import { PromiseHelper } from '../../external/microsoft-cognitiveservices-speech-sdk/distrib/lib/src/common/Promise';
+import { Stream } from '../../external/microsoft-cognitiveservices-speech-sdk/distrib/lib/src/common/Stream';
+
+const CHUNK_SIZE = 4096;
+
+class QueuedArrayBufferAudioSource {
+  constructor(audioFormat, audioSourceId = createNoDashGuid()) {
+    this._audioFormat = audioFormat;
+    this._queue = [];
+    this._id = audioSourceId;
+    this._streams = {};
+
+    this.onEvent = event => {
+      this._events.onEvent(event);
+      Events.instance.onEvent(event);
+    };
+
+    this._events = new EventSource();
+
+    this.attach = this.attach.bind(this);
+    this.detach = this.detach.bind(this);
+    this.id = this.id.bind(this);
+    this.push = this.push.bind(this);
+    this.turnOff = this.turnOff.bind(this);
+    this.turnOn = this.turnOn.bind(this);
+    this.upload = this.upload.bind(this);
```

Comment:
You can do without these if you rewrite the functions slightly - see below comment for an example.